mmdet/models/utils/transformer.py

```diff
@@ -0,0 +1,730 @@
+        else:
+            encoder_order = ('selfattn', 'norm', 'ffn', 'norm')
+            decoder_order = ('selfattn', 'norm', 'multiheadattn', 'norm',
+                             'ffn', 'norm')
+        self.encoder = TransformerEncoder(num_encoder_layers, embed_dims,
+                                          num_heads, feedforward_channels,
+                                          dropout, encoder_order, act_cfg,
+                                          norm_cfg, num_fcs)
+        self.decoder = TransformerDecoder(num_decoder_layers, embed_dims,
+                                          num_heads, feedforward_channels,
+                                          dropout, decoder_order, act_cfg,
+                                          norm_cfg, num_fcs,
+                                          return_intermediate_dec)
+        # follow the official DETR to init parameters
+        self.init_weights()
+
+    def init_weights(self, distribution='uniform'):
+        for m in self.modules():
+            if hasattr(m, 'weight') and m.weight.dim() > 1:
+                xavier_init(m, distribution=distribution)
+
+    def forward(self, x, mask, query_embed, pos_embed):
+        """Forward function for `Transformer`.
+
+        Args:
+            x (Tensor): Input query with shape [bs,c,h,w] where
+                c=embed_dims.
+            mask (Tensor): The key_padding_mask used for encoder and decoder,
+                with shape [bs,h,w].
+            query_embed (Tensor): The query embedding for decoder, with shape
+                [num_query,c].
+            pos_embed (Tensor): The positional encoding for encoder and
+                decoder, with the same shape as `x`.
+
+        Returns:
+            hs (Tensor): Output from decoder. If return_intermediate_dec
+                is True output has shape [num_dec_layers,bs,num_query,
+                embed_dims], else has shape [1,bs,num_query,embed_dims].
+            memory (Tensor): Output results from encoder, with shape
+                [bs,embed_dims,h,w].
+        """
+        bs, c, h, w = x.shape
+        x = x.flatten(2).permute(2, 0, 1)  # [bs,c,h,w] -> [h*w,bs,c]
+        pos_embed = pos_embed.flatten(2).permute(2, 0, 1)
+        query_embed = query_embed.unsqueeze(1).repeat(
+            1, bs, 1)  # [num_query,dim] -> [num_query,bs,dim]
+        mask = mask.flatten(1)  # [bs,h,w] -> [bs,h*w]
+        memory = self.encoder(
+            x, pos=pos_embed, attn_mask=None, key_padding_mask=mask)
+        target = torch.zeros_like(query_embed)
+        # hs: [num_layers,num_query,bs,dim]
+        hs = self.decoder(
+            target,
+            memory,
+            memory_pos=pos_embed,
+            query_pos=query_embed,
+            memory_attn_mask=None,
+            target_attn_mask=None,
+            memory_key_padding_mask=mask,
+            target_key_padding_mask=None)
+        return hs.transpose(1, 2), memory.permute(1, 2, 0).reshape(bs, c, h, w)
```

Comment:
Put these permute and reshape operations into the code block rather than in return